types.d.ts

```diff
@@ -9,14 +9,24 @@ export type RESERVED_7 = 7;
 export type RESERVED_8 = 8;
 export type RESERVED_9 = 9;
 
+export type CallbagArgs<I, O> =
+  // handshake:
+  | [type: START, talkback: Callbag<O, I>]
+  // data from source:
+  | [type: DATA, data: I]
+  // pull request:
+  | [type: DATA]
+  // error:
+  | [type: END, error: unknown]
+  // end without error:
+  | [type: END, _?: undefined]
```

Comment:
Why are the two error cases not one?

Reply:
Because this is stricter. If you specify a single argument with type `END`, or two arguments where second one is `undefined`, TypeScript will infer the second argument to be `undefined`. However if there is anything present in the second argument, it will be considered `unknown` (which is anything that's not `undefined`).

Reply:
I guess the only difference is that this refinement will work correctly:

```
if (args[0] === END && args.length === 1) {
  const arg1: undefined = args[1]
}
```

Whereas if we used `[type: END, error?: unknown]` instead of both, then `args[1]` will always be `unknown`.

I realize it's not a particularly useful refinement, cause you probably wanna check `args[1]` for presence of the error, but it's there just for the sake of completeness. Plus, I like the fact that all the possible states are represented as a separate tuple.